src/CommandingOfficers/Tech.java

```diff
@@ -0,0 +1,452 @@
+
+    return events;
+  }
+
+  /** Heal all units by the specified amount, allowing HP>10, and provide a buff. */
+  private static class OverchargeAbility extends CommanderAbility
+  {
+    private CODamageModifier damageBuff = null;
+    private CODefenseModifier defenseBuff = null;
+    private int healAmount;
+
+    public OverchargeAbility(Commander commander, String abilityName, double abilityCost, int buff, int healAmt)
+    {
+      super(commander, abilityName, abilityCost);
+
+      // Create COModifiers that we can apply when needed.
+      damageBuff = new CODamageModifier(buff);
+      defenseBuff = new CODefenseModifier(buff);
+      healAmount = healAmt;
+
+      // Only mechanical units get the firepower boost.
+      long exclude = UnitModel.TROOP;
+      ArrayList<UnitModel> models = commander.getAllModels(Tech.mechanicalUnits, true, exclude);
+      for( UnitModel m : models )
+        damageBuff.addApplicableUnitModel(m);
+
+      AIFlags = PHASE_TURN_START | PHASE_TURN_END;
+    }
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    protected void perform(MapMaster gameMap)
+    {
+      // Bump up our power level.
+      myCommander.addCOModifier(damageBuff);
+      myCommander.addCOModifier(defenseBuff);
+
+      // Overcharge
+      for(Unit u: myCommander.units)
+      {
+        // Only mechanical units get overcharged.
+        if( u.model.isNone(UnitModel.TROOP) )
+        u.alterHP(healAmount, true);
+      }
+    }
+  }
+
+  /** Drop a MechWarrior onto the most contested point on the battlefront. */
+  private static class TechdropAbility extends CommanderAbility
+  {
+    private static final long serialVersionUID = 1L;
+
+    private CODamageModifier damageBuff = null;
+    private CODefenseModifier defenseBuff = null;
+    private int dropRange;
+    private int numDrops;
+
+    GameEventQueue abilityEvents;
+
+    TechdropAbility(Commander commander, String abilityName, double abilityCost, int buff, int num, int abilityRange)
+    {
+      super(commander, abilityName, abilityCost);
+
+      // Create COModifiers that we can apply when needed.
+      damageBuff = new CODamageModifier(buff);
+      defenseBuff = new CODefenseModifier(buff);
+      numDrops = num;
+      dropRange = abilityRange;
+      abilityEvents = new GameEventQueue();
+
+      // Only mechanical units get the firepower boost.
+      long exclude = UnitModel.TROOP;
+      ArrayList<UnitModel> models = commander.getAllModels(Tech.mechanicalUnits, true, exclude);
+      for( UnitModel m : models )
+        damageBuff.addApplicableUnitModel(m);
+    }
+
+    @Override
+    protected void perform(MapMaster gameMap)
+    {
+      // Bump up our power level.
+      myCommander.addCOModifier(damageBuff);
+      myCommander.addCOModifier(defenseBuff);
+
+      // Perform any events we generated in getEvents().
+      while(!abilityEvents.isEmpty())
+      {
+        abilityEvents.poll().performEvent(gameMap);
+      }
+    }
```

Comment:
I still think it'd be better if `perform()` just returned a `GameEventQueue`. It makes sense from an animation standpoint and a replays/netplay perspective.

Reply:
And that event queue would have to be passed all the way back up to `MapController.animationEnded`? I guess there's no reason we couldn't do that.
Wouldn't `MapController` just have to call `performEvent` on each event in the returned queue anyway, before publishing it to the animator?
I could see it being useful to allow events to create/return sub-events (e.g. `TeleportEvent` can result in a `UnitDieEvent` and possible a `CommanderDefeatEvent`), but I'm sure there would be other ways to handle that case.

Reply:
> And that event queue would have to be passed all the way back up to `MapController.animationEnded`? I guess there's no reason we couldn't do that.

No? We could just have the ability action call `perform()` directly and return the events itself.
We could support both the existing method and the new scheme by making the "official" interface be `getEvents()`, and have that default to producing our current ability event that calls `perform()`.

> Wouldn't `MapController` just have to call `performEvent` on each event in the returned queue anyway, before publishing it to the animator?

Yes? Unless I'm missing something (which I probably am, seeing as I'm not even close to finished reviewing), you're just executing the events and not feeding them to `MapController` to be animated at all.

> I could see it being useful to allow events to create/return sub-events (e.g. `TeleportEvent` can result in a `UnitDieEvent` and possible a `CommanderDefeatEvent`), but I'm sure there would be other ways to handle that case.

We already have a pattern for handling that: the `GameAction` figures out if any of those are needed, and puts them all in the queue in sequence.
There's probably an argument to be made that e.g. a `TeleportEvent` might want to integrate the obstacle's death into the teleport animation itself, so maybe we have good reasons for supporting both patterns?

Reply:
> No? We could just have the ability action call perform() directly and return the events itself.

Yeah, we could make abilities return all/most effects as events and call `getEvents()` in `AbilityAction.getEvents()` instead of just returning a single `AbilityEvent`. I've no problem with that; it would probably be better than my current approach of fetching/handling those extra events underneath `AbilityEvent`.
We can't do this by changing `perform()` though, since it is only called after the animation is over. We could move all functionality into events though, and it would effectively go away.

> Wouldn't MapController just have to call performEvent on each event in the returned queue anyway

I got the order backward here. Events are always animated before being executed and published.

> We already have a pattern for handling that: the GameAction figures out

That allows Actions to generate sub-events, not events to generate sub-events. Right now for example, there is no `TeleportAction`, but `TeleportEvent` is generating the `UnitDieEvent` and `CommanderDefeatEvent` if needed. It keeps track of the sub-events so it can send them to the listeners when it sends itself.
Hm. Maybe I should just create a TeleportAction and have `TechDrop` use that.

> both patterns

To be explicit, they are: requiring all events via `getEvents()` vs. allowing events to generate sub-events at execution time?

Reply:
> Yeah, we could make abilities return all/most effects as events and call `getEvents()` in `AbilityAction.getEvents()` instead of just returning a single `AbilityEvent`.

Right, and we could combine this with making the *default* implementation of Ability's `getEvents()` just return an `AbilityEvent`.
This could result in kind of a weird system where some abilities just never call `perform()` while others use it exclusively, but I feel like I might be okay with that.

> That allows Actions to generate sub-events, not events to generate sub-events. Right now for example, there is no `TeleportAction`, but `TeleportEvent` is generating the `UnitDieEvent` and `CommanderDefeatEvent` if needed. It keeps track of the sub-events so it can send them to the listeners when it sends itself.

Right, and to me that feels like it might be unnecessarily complicated.

> Hm. Maybe I should just create a TeleportAction and have `TechDrop` use that.

Sounds good to me. ~~and then we can make a `TeleportActionFactory` and suddenly we have Swap Mech~~

> > both patterns
> 
> To be explicit, they are: requiring all events via `getEvents()` vs. allowing events to generate sub-events at execution time?

Yessir. The argument I can see for the second one is to allow a single animation to integrate all the effects instead of e.g. "mech falls onto screen... \*beat\*... infantry explodes"
There may be a better solution for that too, though.
(The first idea that pops to mind: have a flag on events to make them not animate, so you can have [`TeleportEvent`, `UnitDieEvent`, CODefeat] in your queue, and the death doesn't need to get animated because it's animated as part of `TeleportEvent`)

Reply:
> Right, and to me that feels like it might be unnecessarily complicated.

I can't disagree. This was the first way I got it working, but I'm sure that whole sequence can be streamlined. Should I try to clean that up in this PR, or make a new one?



I've been toying with the idea of allowing animations to overlap - define a "done but still animating" point at which the next animation in the queue can start. This would make e.g. abilities that buff every unit go faster (though they could potentially be combined into one large animation that handles all that).
Completely up for debate though; the "do/don't animate" flag could also work, though I feel like its usage would be a bit niche. I think I may try decomposing `TeleportEvent` further and see how things look then.

Reply:
> I've been toying with the idea of allowing animations to overlap - define a "done but still animating" point at which the next animation in the queue can start. This would make e.g. abilities that buff every unit go faster (though they could potentially be combined into one large animation that handles all that).

I do kind of like this idea, since it does have slightly broader application, I think.
That said, we'd have to have a *list* of active animations, track when all relevant ones are done, and do something about the possibility of multiple animations drawing the same unit (even if it's "don't let it happen").

> Completely up for debate though; the "do/don't animate" flag could also work, though I feel like its usage would be a bit niche. I think I may try decomposing `TeleportEvent` further and see how things look then.

It's for things that have knock-on effects like unit death. One that comes to mind would be Witchfire - have a buff animation (with a special version for unit death) and don't separately animate the death.